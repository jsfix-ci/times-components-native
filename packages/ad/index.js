export { default } from "./src/ad";
