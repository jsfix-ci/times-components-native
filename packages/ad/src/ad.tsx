import React from "react";

import { useIsConnected } from "@times-components-native/utils/src/useIsConnected";
import { useResponsiveContext } from "@times-components-native/responsive";
import { getNarrowArticleBreakpoint } from "@times-components-native/styleguide";

import { getSlotConfig } from "./utils";

import DOMContext from "./dom-context";

interface IAdConfig {
  sectionName: string;
  slug: string;
}

type TSlotName = "ad-section" | "ad-inarticle-mpu";

interface IProps {
  adConfig: IAdConfig;
  baseUrl?: string;
  keyId?: string;
  narrowContent?: boolean;
  slotName: TSlotName;
}

function Ad({
  adConfig,
  baseUrl = "https://www.thetimes.co.uk/",
  keyId = "",
  narrowContent = false,
  slotName,
}: IProps) {
  const { orientation, windowWidth } = useResponsiveContext();
  const isConnected = useIsConnected();

  const config = getSlotConfig(slotName, windowWidth, orientation);

  const sizeProps = {
    height: config.maxSizes.height,
    width: narrowContent
      ? getNarrowArticleBreakpoint(windowWidth).content
      : windowWidth,
  };

  return (
    <>
      {!isConnected ? null : (
        <DOMContext
          baseUrl={baseUrl}
          keyId={keyId}
          sectionName={adConfig.sectionName}
          slug={adConfig.slug}
          slotName={slotName}
          {...sizeProps}
        />
      )}
    </>
  );
}

export default Ad;
