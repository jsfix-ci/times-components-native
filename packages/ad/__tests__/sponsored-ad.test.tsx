import { SponsoredAd } from "@times-components-native/ad";
import TestRenderer from "react-test-renderer";
import React from "react";
import WebView from "react-native-webview";

jest.mock("react-native", () => {
  const rn = jest.requireActual("react-native");
  rn.NativeModules.Linking = { openURL: jest.fn() };
  return rn;
});

describe("SponsoredAd", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("renders Dianomi script tag", () => {
    const ad = TestRenderer.create(<SponsoredAd numberOfAds={4} />);
    const webView = ad.root.findByType(WebView);

    expect(webView.props.source.html).toMatchSnapshot();
  });

  it("renders the require number of ad items", () => {
    const ad = TestRenderer.create(<SponsoredAd numberOfAds={2} />);
    const webView = ad.root.findByType(WebView);

    expect(webView.props.source.html).toMatchSnapshot();
  });
});
