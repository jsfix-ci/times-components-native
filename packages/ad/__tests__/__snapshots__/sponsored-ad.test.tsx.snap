// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SponsoredAd renders Dianomi script tag 1`] = `
"
            <html>
              <body>
                <script type=\\"text/javascript\\" id=\\"dianomi_context_script\\" src=\\"https://www.dianomi.com/js/contextfeed.js\\"></script>
                <div id=\\"dianomi\\" class=\\"dianomi_context\\" data-dianomi-context-id=\\"243\\"></div>
                <script>
                  (function webviewHeightCallbackSetup(_ref) {
  \\"show source\\";

  var window = _ref.window,
      document = _ref.document,
      MutationObserver = _ref.MutationObserver,
      width = _ref.width,
      ratio = _ref.ratio;
  window.ReactNativeWebView.postMessage(JSON.stringify({
    running: \\"error\\"
  }));
  document.addEventListener(\\"DOMContentLoaded\\", function () {
    var elementToObserve = document.getElementById(\\"dianomi\\");
    window.ReactNativeWebView.postMessage(JSON.stringify({
      running: elementToObserve.clientHeight
    }));
    var observer = new MutationObserver(function () {
      try {
        var dianomiAdContainer = document.getElementById(\\"dianomi\\");
        var adContainerClientHeight = dianomiAdContainer.clientHeight;
        var pixelWidthOfScreen = width * ratio;
        var normalisedAdContainerHeight = pixelWidthOfScreen / window.innerWidth * adContainerClientHeight;
        window.ReactNativeWebView.postMessage(JSON.stringify({
          adContainerHeight: normalisedAdContainerHeight
        }));
      } catch (error) {
        window.ReactNativeWebView.postMessage(JSON.stringify({
          adContainerHeight: 0,
          error: \\"error\\"
        }));
      }
    });
    observer.observe(elementToObserve, {
      characterData: true,
      childList: true,
      attributes: true,
      subtree: true
    });
  });
})({window, document, MutationObserver, width:750, ratio:2});
                </script>
              </body>
            </html>
          "
`;

exports[`SponsoredAd renders the require number of ad items 1`] = `
"
            <html>
              <body>
                <script type=\\"text/javascript\\" id=\\"dianomi_context_script\\" src=\\"https://www.dianomi.com/js/contextfeed.js\\"></script>
                <div id=\\"dianomi\\" class=\\"dianomi_context\\" data-dianomi-context-id=\\"250\\"></div>
                <script>
                  (function webviewHeightCallbackSetup(_ref) {
  \\"show source\\";

  var window = _ref.window,
      document = _ref.document,
      MutationObserver = _ref.MutationObserver,
      width = _ref.width,
      ratio = _ref.ratio;
  window.ReactNativeWebView.postMessage(JSON.stringify({
    running: \\"error\\"
  }));
  document.addEventListener(\\"DOMContentLoaded\\", function () {
    var elementToObserve = document.getElementById(\\"dianomi\\");
    window.ReactNativeWebView.postMessage(JSON.stringify({
      running: elementToObserve.clientHeight
    }));
    var observer = new MutationObserver(function () {
      try {
        var dianomiAdContainer = document.getElementById(\\"dianomi\\");
        var adContainerClientHeight = dianomiAdContainer.clientHeight;
        var pixelWidthOfScreen = width * ratio;
        var normalisedAdContainerHeight = pixelWidthOfScreen / window.innerWidth * adContainerClientHeight;
        window.ReactNativeWebView.postMessage(JSON.stringify({
          adContainerHeight: normalisedAdContainerHeight
        }));
      } catch (error) {
        window.ReactNativeWebView.postMessage(JSON.stringify({
          adContainerHeight: 0,
          error: \\"error\\"
        }));
      }
    });
    observer.observe(elementToObserve, {
      characterData: true,
      childList: true,
      attributes: true,
      subtree: true
    });
  });
})({window, document, MutationObserver, width:750, ratio:2});
                </script>
              </body>
            </html>
          "
`;
